### PR TITLE
feat: adapt `eth_sender` tests for running both in Validium mode and Rollup mode

### LINKED ISSUE #118: Adapt tests for testing both with Rollup and Validium mode in `eth_sender` module
## Descripction

Adapt tests for testing both with Rollup and Validium mode (abstract the current logic, create a tester for each mode, and send it to the abstracted logic) for the `eth_sender` tests (`core/lib/zksync_core/src/eth_sender/tests.rs`).

## How to run `eth_sender` tests

```
zk test rust --package zksync_core --lib -j 1 -- eth_sender::tests
```

## Example

**Path to file:** `core/lib/zksync_core/src/eth_sender/tests.rs`
**Test:** `confirm_many`
**Run:** `zk test rust --package zksync_core --lib -j 1 -- eth_sender::tests::confirm_many`
**Test code:**
```rust
// Tests that we send multiple transactions and confirm them all in one iteration.
#[tokio::test]
async fn confirm_many() -> anyhow::Result<()> {
    let connection_pool = ConnectionPool::test_pool().await;
    let l1_batch_commit_data_generator = Arc::new(RollupModeL1BatchCommitDataGenerator {});
    let mut tester = EthSenderTester::new(
        connection_pool,
        vec![10; 100],
        false,
        l1_batch_commit_data_generator.clone(),
    )
    .await;

    let mut hashes = vec![];

    for _ in 0..5 {
        let tx = tester
            .aggregator
            .save_eth_tx(
                &mut tester.conn.access_storage().await.unwrap(),
                &DUMMY_OPERATION,
                true,
            )
            .await?;
        let hash = tester
            .manager
            .send_eth_tx(
                &mut tester.conn.access_storage().await.unwrap(),
                &tx,
                0,
                L1BlockNumber(tester.gateway.block_number("").await?.as_u32()),
            )
            .await?;
        hashes.push(hash);
    }

    // check that we sent something
    assert_eq!(tester.gateway.sent_tx_count(), 5);
    assert_eq!(
        tester
            .storage()
            .await
            .eth_sender_dal()
            .get_inflight_txs()
            .await
            .unwrap()
            .len(),
        5
    );

    for hash in hashes {
        tester
            .gateway
            .execute_tx(hash, true, EthSenderTester::WAIT_CONFIRMATIONS);
    }

    let to_resend = tester
        .manager
        .monitor_inflight_transactions(
            &mut tester.conn.access_storage().await.unwrap(),
            tester.get_block_numbers().await,
        )
        .await?;

    // check that transaction is marked as accepted
    assert_eq!(
        tester
            .storage()
            .await
            .eth_sender_dal()
            .get_inflight_txs()
            .await
            .unwrap()
            .len(),
        0
    );

    // also check that we didn't try to resend it
    assert!(to_resend.is_none());

    Ok(())
}
```

To adapt this test for running both in Rollup and in Validum mode we can do the following:

```rust
async fn _confirm_many(tester: &mut EthSenderTester) -> anyhow::Result<()> {
    let mut hashes = vec![];

    for _ in 0..5 {
        let tx = tester
            .aggregator
            .save_eth_tx(
                &mut tester.conn.access_storage().await.unwrap(),
                &DUMMY_OPERATION,
                true,
            )
            .await?;
        let hash = tester
            .manager
            .send_eth_tx(
                &mut tester.conn.access_storage().await.unwrap(),
                &tx,
                0,
                L1BlockNumber(tester.gateway.block_number("").await?.as_u32()),
            )
            .await?;
        hashes.push(hash);
    }

    // check that we sent something
    assert_eq!(tester.gateway.sent_tx_count(), 5);
    assert_eq!(
        tester
            .storage()
            .await
            .eth_sender_dal()
            .get_inflight_txs()
            .await
            .unwrap()
            .len(),
        5
    );

    for hash in hashes {
        tester
            .gateway
            .execute_tx(hash, true, EthSenderTester::WAIT_CONFIRMATIONS);
    }

    let to_resend = tester
        .manager
        .monitor_inflight_transactions(
            &mut tester.conn.access_storage().await.unwrap(),
            tester.get_block_numbers().await,
        )
        .await?;

    // check that transaction is marked as accepted
    assert_eq!(
        tester
            .storage()
            .await
            .eth_sender_dal()
            .get_inflight_txs()
            .await
            .unwrap()
            .len(),
        0
    );

    // also check that we didn't try to resend it
    assert!(to_resend.is_none());

    Ok(())
}

// Tests that we send multiple transactions and confirm them all in one iteration.
#[tokio::test]
async fn confirm_many() -> anyhow::Result<()> {
    let mut rollup_tester = EthSenderTester::new(
        ConnectionPool::test_pool().await,
        vec![10; 100],
        false,
        Arc::new(RollupModeL1BatchCommitDataGenerator {}),
    )
    .await;
    let mut validium_tester = EthSenderTester::new(
        ConnectionPool::test_pool().await,
        vec![10; 100],
        false,
        Arc::new(ValidiumModeL1BatchCommitDataGenerator {}),
    )
    .await;

    _confirm_many(&mut rollup_tester).await?;
    _confirm_many(&mut validium_tester).await
}
```

**Note:** this is more than an example, this test passes
```shell
    Starting 1 test across 1 binary (227 skipped)
        PASS [   0.310s] zksync_core eth_sender::tests::confirm_many
------------
     Summary [   0.317s] 1 test run: 1 passed, 227 skipped
```

### PR DESCRIPTION
**Pull request by toni-calvin**
## Description
Closes #118 

This Pull Request addresses the need to support and test a new type of SenderTester, specifically the "validium" tester, in addition to the default "rollup" tester. In a previous PR, we introduced different testers, but the default was set to "rollup" without adequate testing for the "validium" tester.

The main objective of this PR is to enhance testing coverage for the "validium" tester. The existing test structure has been adjusted to accept a SenderTest as a parameter, allowing for greater flexibility in testing various tester types. Additionally, new methods have been added to validate the tests specifically for each tester type, ensuring thorough and targeted testing for both "rollup" and "validium" testers.

## Changes Made

- Adjustment of the `l1_batch_commit_generator` to accommodate the unique requirements of the "validium" tester, as it is the only differing aspect between tester types.
- Adaptation of the existing test structure to accept a SenderTest as a parameter.
- Addition of methods that instantiate one of each type of SenderTester (including "rollup" and "validium") and call the existing tests, now with the Tester as a parameter.

## How to Test

- For testing, inside the root folder of the project run: 

`zk test rust --package zksync_core --lib -j 1 -- eth_sender::tests`

### CODE REVIEW

File: core/lib/zksync_core/src/eth_sender/tests.rs [hunk 1]
```diff
@@ -657,59 +703,72 @@ async fn correct_order_for_confirmations() -> anyhow::Result<()> {
...
+    )
+    .await;
+
+    _correct_order_for_confirmations(&mut rollup_tester).await?;
+    _correct_order_for_confirmations(&mut validium_tester).await
+}
+
+async fn _skipped_l1_batch_at_the_start(tester: &mut EthSenderTester) -> anyhow::Result<()> {
+    let l1_batch_commit_data_generator: Arc<dyn L1BatchCommitDataGenerator> =
+        tester.aggregator.get_l1_batch_commit_data_generator();
```
Comment (ilitteri) [inline core/lib/zksync_core/src/eth_sender/tests.rs:728]: Can't we just do this? (We'd get rid of the getter method)
```suggestion
    let l1_batch_commit_data_generator: Arc<dyn L1BatchCommitDataGenerator> =
        tester.aggregator.l1_batch_commit_data_generator.clone();
```

**Comment by ilitteri** [comment]
The base branch should be `feat_validium_pubdata_abstraction`.

File: core/lib/zksync_core/src/eth_sender/tests.rs (hunk 1 above)
Comment (toni-calvin) [reply core/lib/zksync_core/src/eth_sender/tests.rs:728]: In order to do that, we should expose public the aggregator field inside the `EthTxAggregator` and the field `l1_batch_commit_data_generator`.

**Comment by ilitteri** [comment]
Let's update the PR's description

**Review by ilitteri — CHANGES_REQUESTED** [review]
The PR is looking good, great work! Here I leave some comments:

- Let's add a description to the PR, given that there's an issue associated we can mention `Resolves #X` or `Closes #X`. Then if some additional information is needed by the reviewer it'd be good to add it in subsequent sections (like how to run the tests you're changing).
- Let's move the auxiliary functions (the abstractions) to a `test_helpers` module. If I'm not wrong, by doing so we an get rid of the prefix `_` and call the methods like `test_helpers::xxx` after importing the module `test_helpers`.
- Regarding the `get_l1_batch_commit_data_generator` method: I see why you added it, but I think we can avoid needing it (in terms of design it'd be ok to have it as a method for `Aggregator` by maybe not for `EthTxAggregator`, we'll still remove it from both though). As it is only meant for testing purposes we can make the method public for the crate or super if necessary, even better, we can make the struct fields public for the crate or super; this won't be the case either because it'd be more difficult to be accepted in the main PR.
    I saw in the tests that it is only consumed by `commit_l1_batch` to instantiate an `AggregatedOperation::Commit` which instantiates a `CommitBatches` struct that will be the final consumer of `l1_batch_commit_data_generator`. This is great news because this means that the `EthTxAggregator`'s `Aggregator` that is in the `tester` is not the one in charge of creating an `AggregatedOperation::Commit` for later committing, therefore we can pass `l1_batch_commit_data_generator` as a parameter to the functions along with the `tester` and get rid of using the method and roll it back.

**Review by ilitteri — APPROVED** [review]
LGTM! Great work!

**Pull request merged**